File: cvoxelise/MeshWriter.cpp
#include "MeshWriter.h"

void write_inp(std::string FILENAME, int type, py::array_t<double> nodes, py::array_t<int> elements) {
	int ele_offset = 1;
	FILE* fp;
	fp = fopen(FILENAME.c_str(), "w+");
	fprintf(fp, "**\n*Node\n");
	for (int i = 0; i < nodes.shape(0); i++)
		fprintf(fp, "%d,   %.6f, %.6f, %.6f\n", i + 1,
			nodes.mutable_at(i, 0),
			nodes.mutable_at(i, 1),
			nodes.mutable_at(i, 2));
	fprintf(fp, "\n*Element, type=%s\n", VTK_TO_ABAQUS_TYPE.at(type).c_str());
	for (int j = 0; j < elements.shape(0); j++)
		fprintf(fp, "%d,   %d, %d, %d, %d, %d, %d, %d, %d\n", j + 1,
			elements.mutable_at(j, 0) + ele_offset,
			elements.mutable_at(j, 1) + ele_offset,
			elements.mutable_at(j, 2) + ele_offset,
			elements.mutable_at(j, 3) + ele_offset,
			elements.mutable_at(j, 4) + ele_offset,
			elements.mutable_at(j, 5) + ele_offset,
			elements.mutable_at(j, 6) + ele_offset,
			elements.mutable_at(j, 7) + ele_offset);
	fclose(fp);
}

void write_vtk(std::string FILENAME, int type, py::array_t<double> nodes, py::array_t<int> elements) {
	FILE* fp;
	fp = fopen(FILENAME.c_str(), "w+");
	fprintf(fp, 
		"# vtk DataFile Version 3.0\n"
		"Volume Mesh\n"
		"ASCII\n"
		"DATASET UNSTRUCTURED_GRID\n");

	fprintf(fp, "POINTS  %d  float\n", int(nodes.shape(0)) );
	for (int i = 0; i < nodes.shape(0); i++)
		fprintf(fp, "%.9f  %.9f  %.9f\n",
			nodes.mutable_at(i, 0),
			nodes.mutable_at(i, 1),
			nodes.mutable_at(i, 2));
	
	fprintf(fp, "CELLS  %d  %d\n", int(elements.shape(0)), int(elements.size() + elements.shape(0)));
	switch (elements.shape(1))
	{
	case 4:
		for (int j = 0; j < elements.shape(0); j++)
			fprintf(fp, "%d %d %d %d %d\n", 4,
				elements.mutable_at(j, 0) ,
				elements.mutable_at(j, 1) ,
				elements.mutable_at(j, 2) ,
				elements.mutable_at(j, 3) );
		break;
	case 8:
		for (int j = 0; j < elements.shape(0); j++)
			fprintf(fp, "%d %d %d %d %d %d %d %d %d\n", 8,
				elements.mutable_at(j, 0) ,
				elements.mutable_at(j, 1) ,
				elements.mutable_at(j, 2) ,
				elements.mutable_at(j, 3) ,
				elements.mutable_at(j, 4) ,
				elements.mutable_at(j, 5) ,
				elements.mutable_at(j, 6) ,
				elements.mutable_at(j, 7) );
		break;
	}

	fprintf(fp, "CELL_TYPES  %d\n", int(elements.shape(0)));
	for (int j = 0; j < elements.shape(0); j++)
		fprintf(fp, "%d\n", type);

	fclose(fp);
}

void write_vtu(std::string FILENAME, int type, py::array_t<double> nodes, py::array_t<int> elements) {
	FILE* fp;
	int header,hlen;
	fp = fopen(FILENAME.c_str(), "w+");
	fprintf(fp,
		"<?xml version = \"1.0\"?>\n"
		"<VTKFile type = \"UnstructuredGrid\" version = \"0.1\" byte_order = \"LittleEndian\">\n"
		"<UnstructuredGrid>\n"
		"<Piece NumberOfPoints = \"%d\" NumberOfCells = \"%d\"> \n", int(nodes.shape(0)), int(elements.shape(0)));

	fprintf(fp, 
		"<Points>\n"
		"<DataArray type = \"Float32\" Name = \"Points\" NumberOfComponents = \"3\" format = \"binary\">\n");
	
	py::array_t<float> newnodes = nodes.cast<py::array_t<float>>();

	char* encoding = base64((char*)newnodes.data(), sizeof(float) * nodes.size(), &header);
	fprintf(fp, base64((char*)&header, sizeof(int), &hlen));
	fprintf(fp, encoding);

	fprintf(fp, 
		"\n"
		"</DataArray>\n"
		"</Points>\n"
		"<Cells>\n"
		"<DataArray type = \"Int32\" Name = \"connectivity\" format = \"binary\">\n");

	encoding = base64((char*)elements.data(), sizeof(int) * elements.size(), &header);
	fprintf(fp, base64((char*)&header, sizeof(int), &hlen));
	fprintf(fp, encoding);

	py::array_t<int> offsets = py::array_t<int>(elements.shape(0));
	py::array_t<int> types = py::array_t<int>(elements.shape(0));

#pragma omp parallel for
	for (int i = 0; i < elements.shape(0); i++) {
		offsets.mutable_at(i) = int(elements.shape(1)) * (1 + i);
		types.mutable_at(i) = type;
	}
	
	fprintf(fp,
		"\n"
		"</DataArray>\n"
		"<DataArray type = \"Int32\" Name = \"offsets\" format = \"binary\">\n");

	encoding = base64((char*)offsets.data(), sizeof(int) * offsets.size(), &header);
	fprintf(fp, base64((char*)&header, sizeof(int), &hlen));
	fprintf(fp, encoding);

	fprintf(fp,
		"\n"
		"</DataArray>\n"
		"<DataArray type = \"Int32\" Name = \"types\" format = \"binary\">\n");

	encoding = base64((char*)types.data(), sizeof(int) * types.size(), &header);
	fprintf(fp, base64((char*)&header, sizeof(int), &hlen));
	fprintf(fp, encoding);

	fprintf(fp,
		"\n"
		"</DataArray>\n"
		"</Cells>\n"
		"</Piece>\n"
		"</UnstructuredGrid>\n"
		"</VTKFile>\n");

	fclose(fp);
	return;
}

File: meshWriter.py
import numpy as np
import base64

VTK_TO_ABAQUS_TYPE = {
    8: "S4R",
    9: "S4R",
    10: "C3D4",
    11: "C3D8R",
    12: "C3D8R"
}

def write_inp(FILENAME:str, type:int, nodes:np.ndarray, elements:np.ndarray):
    ele_offset = 1
    with open(FILENAME, "wt") as f:
        f.write("*HEADING\n")
        f.write('\n*Node\n')
        fmt = ", ".join(["{}"] + ["{:.9f}"] * nodes.shape[1]) + "\n"
        for k, x in enumerate(nodes):
            f.write(fmt.format(k + 1, *x))
        
        ele_nod = elements + ele_offset
        f.write('\n*Element, type={}\n'.format(VTK_TO_ABAQUS_TYPE[type]))

        for e in range(1, elements.shape[0] + 1):
            f.write('{:d},  {:d}, {:d}, {:d}, {:d}, {:d}, {:d}, {:d}, {:d}\n'.format(e,*ele_nod[e-1,:]))
    return



def write_vtk(FILENAME:str, type:int, nodes:np.ndarray, elements:np.ndarray):
    with open(FILENAME, "wt") as f:
        f.write(
            "# vtk DataFile Version 5.1\n"
            "Volume Mesh\n"
            "ASCII\n"
            "DATASET UNSTRUCTURED_GRID\n")

        f.write("POINTS  {:d}  float\n".format(nodes.shape[0]) )
        nodes.tofile(f, sep=" ")
        f.write("\n")

        f.write("CELLS {:d} {:d}\n".format(elements.shape[0]+1, elements.size))
        f.write("OFFSETS vtktypeint64\n")
        offsets = np.arange(0, elements.size+1, elements.shape[1], dtype=int)
        offsets.tofile(f, sep="\n")
        f.write("\n")

        f.write("CONNECTIVITY vtktypeint64\n")
        elements.tofile(f, sep="\n")
        f.write("\n")

        f.write("CELL_TYPES  {:d}\n".format(elements.shape[0]))
        np.full(elements.shape[0], type).tofile(f, sep="\n")
        f.write("\n")
    return



def write_vtu(FILENAME:str, type:int, nodes:np.ndarray, elements:np.ndarray):
    with open(FILENAME, "wt") as f:
        f.write(
            "<?xml version = \"1.0\"?>\n"
            "<VTKFile type = \"UnstructuredGrid\" version = \"0.1\" byte_order = \"LittleEndian\">\n"
            "<UnstructuredGrid>\n"
            "<Piece NumberOfPoints = \"{:d}\" NumberOfCells = \"{:d}\"> \n".format(nodes.shape[0], elements.shape[0])  )

        f.write(
            "<Points>\n"
            "<DataArray type = \"Float32\" Name = \"Points\" NumberOfComponents = \"3\" format = \"binary\">\n")
        data_bytes = nodes.astype(np.float32).tobytes()
        header = np.array(len(data_bytes), dtype=np.uint)
        f.write(base64.b64encode(header.tobytes() + data_bytes).decode())

        f.write("\n"
            "</DataArray>\n"
            "</Points>\n"
            "<Cells>\n"
            "<DataArray type = \"Int32\" Name = \"connectivity\" format = \"binary\">\n")
        data_bytes = elements.astype(np.int32).tobytes()
        header = np.array(len(data_bytes), dtype=int)
        f.write(base64.b64encode(header.tobytes() + data_bytes).decode())

        f.write("\n"
            "</DataArray>\n"
            "<DataArray type = \"Int32\" Name = \"offsets\" format = \"binary\">\n")
        offsets = np.arange(elements.shape[1], elements.size+1, elements.shape[1], dtype=np.int32)
        data_bytes = offsets.tobytes()
        header = np.array(len(data_bytes), dtype=int)
        f.write(base64.b64encode(header.tobytes() + data_bytes).decode())

        f.write("\n"
            "</DataArray>\n"
            "<DataArray type = \"Int32\" Name = \"types\" format = \"binary\">\n")
        types = np.full(elements.shape[0], type).astype(np.int32)
        data_bytes = types.tobytes()
        header = np.array(len(data_bytes), dtype=int)
        f.write(base64.b64encode(header.tobytes() + data_bytes).decode())

        f.write("\n"
            "</DataArray>\n"
            "</Cells>\n"
            "</Piece>\n"
            "</UnstructuredGrid>\n"
            "</VTKFile>\n")
    return



if __name__=='__main__':
    print("# vtk DataFile Version 5.1\n"
        "Volume Mesh\n"
        "ASCII\n"
        "DATASET UNSTRUCTURED_GRID\n")
    data_bytes = np.full(100,33333).tobytes()
    header = np.array(len(data_bytes), dtype=np.float64)
    print(int(33333).to_bytes(length=4,byteorder='little'))
    print(header.tobytes())
    print(base64.b64encode(header.tobytes() + data_bytes).decode())